File: MinStack/MinStack.cpp
class MinStack {
public:
    vector<int> elements;
    vector<int> minIndexes;
    void push(int x) {
        if (elements.size() == 0) {
            elements.push_back(x);
            minIndexes.push_back(0);
        } else {
            elements.push_back(x);
            if (x < elements[minIndexes.back()]) {
                minIndexes.push_back(elements.size()-1);
            }
        }
    }

    void pop() {
        if (minIndexes.back() == elements.size()-1) {
            minIndexes.pop_back();
        }
        elements.pop_back();
    }

    int top() {
        return elements.back();
    }

    int getMin() {
        return elements[minIndexes.back()];
    }
};

File: MinStack/MinStack.js
/**
 * @constructor
 */
var MinStack = function() {
  this.elements = [];
  this.minIndexes = [];
};

/**
 * @param {number} x
 * @returns {void}
 */
MinStack.prototype.push = function(x) {
  this.elements.push(x);
  if (this.minIndexes.length == 0 ||
      x < this.elements[this.minIndexes[this.minIndexes.length - 1]]) {
    this.minIndexes.push(this.elements.length - 1);
  }
};

/**
 * @returns {void}
 */
MinStack.prototype.pop = function() {
  if (this.minIndexes[this.minIndexes.length - 1] === this.elements.length - 1) {
    this.minIndexes.pop();
  }
  this.elements.pop();
};

/**
 * @returns {number}
 */
MinStack.prototype.top = function() {
  return this.elements[this.elements.length - 1];
};

/**
 * @returns {number}
 */
MinStack.prototype.getMin = function() {
  return this.elements[this.minIndexes[this.minIndexes.length - 1]];
};
